---
title: "Zunqiu_Wang_ngs_wk13_hw"
author: "Zunqiu Wang"
date: "4/26/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Task 1: Understanding k-mer profiles

Report the profile in the form of a two column table for your answer [ 1 point ].

```{r,echo=FALSE,eval=TRUE}
library(gt)
library(dplyr)

tb1 <- read.table("/Users/zunqiuwang/Dropbox/NYU_Bioinfo/NGS/WK13/kmer_table.csv", header = T)
tb1 %>% gt() %>% tab_header(title="K-mer Frequency Table")
```

Task 2: Generate k-mer profile with Jellyfish

Q2.1. GenomeScope requires the read length for the fastq input file. What is the read length for the concatenated fastq for Arabidopsis [ 1 point ]?
  
+ 90

Q2.2 Copy your slurm script with all commands into your homework document for your answer [ 2 points ]

```{bash, echo=TRUE, eval=FALSE}
#!/bin/bash
#
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --time=8:00:00
#SBATCH --mem=4GB
#SBATCH --job-name=kmer_count
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=zw2946@nyu.edu

module purge

module load jellyfish/2.3.0

jellyfish count -C -m 21 -s 1000000000 -t 10 cat.fq -o reads.jf

jellyfish histo -t 10 reads.jf > reads.histo
```

# Task 3: Reference-free estimation of heterozygosity, genome size and repeat content

Q3.1 Report the settings (kmer length, read length, and max kmer coverage ) you input on the GenomeScope website [ 1 point ].

+ kmer length: 19 (date palm GenomeScope results that were given to compare with is set k=19) but optically is k=21
+ read length: 90
+ max kmer coverage: 1000

Q3.2. Report your GenomeScope outputs [ 1 point ] (1) Paste the GenomeScope Results table output for Arabidopsis into your homework document (2) Include the two kmer profile plots generated by GenomeScope. Download images or take a screen shot of the output plots and include in your report. If you dont know how to incorporate in a markdown report you may upload screenshots separately. If uploading separately, please include your name in the filename

+ Results

GenomeScope version 1.0
k = 19

property                      min               max               
Heterozygosity                0.108366%         0.112945%         
Genome Haploid Length         124,192,988 bp    124,271,914 bp    
Genome Repeat Length          22,216,876 bp     22,230,995 bp     
Genome Unique Length          101,976,111 bp    102,040,918 bp    
Model Fit                     97.3916%          99.6102%          
Read Error Rate               0.120693%         0.120693%  

<center>

![kmer freq plot](plot.png){width=85%}

</center>

<center>

![kmer freq log plot](plot.log.png){width=85%}

</center>


Q3.3a. Does the kmer-profiling estimates of heterozygosity (the proportion of heterozygous sites) for Arabidopsis and date palm match this expectation? [ 1 point ].

+ yes, Arabidopsis has no visible heterozygosity peak (first vertical dotted line) but one huge homozygosity peak (second vertical dotted line) in both plot, which indicates of low heterozygosity. Statistic output for het rate is 0.111%.
+ yes, date palm has clear heterozygosity peak (first vertical dotted line) but one huge homozygosity peak in both plot, which indicates of comparatively high heterozygosity. Statistic output for het rate is 1.25%.

Q3.3b. From left to right, which of the peaks in the output profile plot for date palm is the homozygous peak, the heterozygous peak, the error peak? Why is the heterozygous peak expected to be half of the homozygous peak (in terms of number of occurences or “Coverage”)? [ 1 points ]

+ the error peak along the red line, black curve at intersection of first vertical dotted line is heterozygous peak, black curve at intersection of second vertical dotted line is homozygous peak
+ homozygous peak is from both strands while heterozygous peak accounts for each strand so coverage is half

Q3.4. A published assembly for date palm has a haploid length of 772 Mb, but may be larger (Hazzouri et al. 2019). The Arabidopsis thaliana assembly is much smaller (~135 Mb). Summarize the haploid genome size estimates of GenomeScope for these two species? Then speculate as to why GenomeScope estimates might differ from the genome assembly sizes for both species [ 1 point ].

+ date palm: 355,424,186
+ Arabidopsis thaliana: 124,271,914
+ short reads failed to map over the most heterozygous and repeated regions. Maybe one genome is significantly larger than the other, because its estimate is based on those confidently aligned regions of the genomes

Q3.5. Hypothetically-speaking, how would you expect the k-mer profile to differ between two fastqs from the same sample with a 2-fold difference in sequencing error rate? [ 1 point ]

+ red line is shifted towards right, and the area under the curve is larger(increased proportion)
















